DataFormats/L1TrackTrigger/interface/TTTrack.h

```diff
@@ -0,0 +1,492 @@
+  }
+
+  if (nPar==5) {
+    return theChi25Par;
+  }
+
+  return 0.0;
+
+}
+
+
+
+/// Chi2 reduced
+template< typename T >
+double TTTrack< T >::getChi2Red(unsigned int nPar) const
+{
+
+  if (!checkValidArgs(nPar)) return 0.0;
+
+  if (nPar==4) {
+    return theChi24Par/( 2*theStubRefs.size() - 4 );
+  }
+
+  if (nPar==5) {
+    return theChi25Par/( 2*theStubRefs.size() - 5 );
+  }
+
+  return 0.0;
+
+}
+
+
+/// StubPtConsistency 
+template< typename T >
+void TTTrack< T >::setStubPtConsistency(double aStubPtConsistency, unsigned int nPar) {
+
+  if (!checkValidArgsForSet(nPar)) return;
+
+  if (nPar==4) {
+    valid4ParFit = true;
+    theStubPtConsistency4Par=aStubPtConsistency;
+  }
+
+  if (nPar==5) {
+    valid5ParFit = true;
+    theStubPtConsistency5Par=aStubPtConsistency;
+  }
+
+  return;
+
+}
+
+
+
+/// StubPtConsistency 
+template< typename T >
+double TTTrack< T >::getStubPtConsistency(unsigned int nPar) const
+{
+
+  if (!checkValidArgs(nPar)) return 0.0;
+
+  if (nPar==4) {
+    return theStubPtConsistency4Par;
+  }
+
+  if (nPar==5) {
+    return theStubPtConsistency5Par;
+  }
+
+  return 0.0;
+
+}
+
+
+
+
+
+/// Duplicate identification
+template< typename T>
+bool TTTrack< T >::isTheSameAs( TTTrack< T > aTrack ) const
+{
+  /// Take the other stubs
+  std::vector< edm::Ref< edmNew::DetSetVector< TTStub< T > >, TTStub< T > > > otherStubRefs = aTrack.getStubRefs();
+
+  /// Count shared stubs
+  unsigned int nShared = 0;
+  for ( unsigned int i = 0; i < theStubRefs.size() && nShared < 2; i++)
+  {
+    for ( unsigned int j = 0; j < otherStubRefs.size() && nShared < 2; j++)
+    {
+      if ( theStubRefs.at(i) == otherStubRefs.at(j) )
+      {
+        nShared++;
+      }
+    }
+  }
+
+  /// Same track if 2 shared stubs
+  return ( nShared > 1 );
+}
+
+/// Quality criteria: does it have a Stub in a specific Barrel Layer?
+template< typename T >
+bool TTTrack< T >::hasStubInBarrel( unsigned int aLayer ) const
+{
+  for ( unsigned int i = 0; i < theStubRefs.size(); i++)
+  {
+    DetId detid =  theStubRefs.at(i)->getDetId();
+    if ( detid.subdetId()==StripSubdetector::TOB )
+    {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+
+/// Information
+template< typename T >
+std::string TTTrack< T >::print( unsigned int i ) const
+{
+  std::string padding("");
+  for ( unsigned int j = 0; j != i; ++j )
+  {
+    padding+="\t";
+  }
+
+  std::stringstream output;
+  output<<padding<<"TTTrack:\n";
+  padding+='\t';
+  output << '\n';
+  unsigned int iStub = 0;
+
+  typename std::vector< edm::Ref< edmNew::DetSetVector< TTStub< T > >, TTStub< T > > >::const_iterator stubIter;
+  for ( stubIter = theStubRefs.begin();
+        stubIter!= theStubRefs.end();
+        ++stubIter )
+  {
+    output << padding << "stub: " << iStub++ << ", DetId: " << ((*stubIter)->getDetId()).rawId() << '\n';
+  }
+
+  return output.str();
+}
+
+
+template< typename T >
+bool TTTrack< T >::checkValidArgs(unsigned int nPar) const {
```

Comment:
I wonder about the logic flow of this part of the code. Would it be preferable to throw an exception here, rather than letting the algorithm progress with default values?